src/fileio/dataformat.cxx

```diff
@@ -46,9 +46,9 @@ void DataFormat::writeFieldAttributes(const std::string& name, const FieldPerp&
 
   auto& fieldmesh = *f.getMesh();
   int yindex = f.getIndex();
-  if (yindex >= 0 and yindex < fieldmesh.LocalNy) {
+  if (yindex >= fieldmesh.ystart and yindex <= fieldmesh.yend) {
```

Comment:
What are valid values for `yindex` and `yindex_global`? `0 < yindex < local_ny` and `0 < yindex_global < ny - 1`? Or `ystart <= yindex <= yend` and `myg <= yindex_global < ny - myg`? (that is, can you have `FieldPerp`s in the boundaries?)

I feel like either the form I've changed it to, where only interior points get saved, or the previous form, where points on this process get saved must be correct, but then one of `test-io` or `test-restart-io` ends up failing.

The problem might be that `collect` assumes that a `FieldPerp` will have `yindex_global == -1` in all but one file. But using `getLocalYIndex`, we can easily end up saving it on multiple processors, due to the guard cells.

For example, in `test-io` we check that reading and then writing a restart file is correct for a `FieldPerp` with `yindex_global = 11`. The grid has `ny = 16` and `myg = 2`, and the failing test is on 4 cores with `NYPE = 4`, so the `LocalNy = 7` (with 3 interior points).

With this set up, we have the following values for `getLocalYIndex(yindex_global)`:

| Processor | Local index |
|------------|------------------|
| 0 | 11 |
| 1 | 8 |
| 2 | 5 |
| 3 | 2 | 

So processors 2 & 3 both save the same point using the old form of the conditional here, which triggers an `assert` in `collect`.

On the other hand, `test-restart-io` tries to check with two `FieldPerp`s, one with `yindex_global = 0`, the other with `16` (note I think this should be `ny - 1 = 15` anyway). These are both in the boundary, so fail with the new form of the conditional.

Do we ever want to read/write `FieldPerp`s in boundaries (and not guards)? If no, then I think this is easily solved by just changing the numbers in `test-restart-io`. Otherwise, I guess we'll need to check `firstY/lastY` and modify the conditional accordingly? If we want to be able to save in _guards_ as well, `collect` will need to be modified.

Now that I've written all this out, I suspect we probably do want them in boundaries!

Note, I have no idea how xBOUT handles this.

Reply:
Sorry, I meant to think about this earlier but got distracted by something else...

We definitely want `FieldPerp`s in boundaries. For example, to save an upper-target boundary condition I'd want to save a `CELL_YLOW` `FieldPerp` at `yend+1`.

Short version of what I say below: I think we should use `if (yindex >= 0 and yindex < fieldmesh.LocalNy)` to test for validity, but only add `fperp2` to `dump` in the tests if its index is on a non-guard point. We allow users to save a `FieldPerp` in guard cells if they want to, but they are responsible for the errors from `collect()` if their `FieldPerp`s clash or are inconsistent.

#### long version

Not sure about guard cells. I'd be tempted to allow this to be done - it might be useful for debugging or something, but if a model saves `FieldPerp`s from different processors at the same global yindex (as defined by `getGlobalYIndex()`, then `collect` should give an error and the user can sort it out by hand.
Actually, I think that ideally a given `FieldPerp` (at least if it's going to be saved and not just used as a temporary, e.g. inside `Laplacian::solve()`) should only be allowed to exist with a single global y-index, and on a single set of processors with the same `PE_YIND`. In this view, we shouldn't actually call for example `dump.add(fperp2, "fperp2", false);` on all the processors, we should only call it on the ones where `fperp2.getIndex()` is not in a guard cell. Unfortunately I don't see any way to enforce this, but we could document it and raise a helpful error in `collect()`.

Sorry, the previous paragraph wasn't very well written... The idea I had in mind in saving with `yindex_global` was that if we're going to have a physical quantity represented by a `FieldPerp` then it should be at a well-defined and unique physical position (with a global y-index in either grid or boundary cells, but guard cells would not make sense). Only then would it be allowed to save it. Other processors where the physical field is not present should not really allocate it or add it to `dump`.

If we do want to check for boundaries explicitly, we should use
https://github.com/boutproject/BOUT-dev/blob/b810b18886c9f93d6399eb247e6e6cab681f8f06/include/bout/mesh.hxx#L457-L458
because `firstY()` and `lastY()` do not account for an upper target (double null configuration). Although their docstrings suggest that they do
https://github.com/boutproject/BOUT-dev/blob/b810b18886c9f93d6399eb247e6e6cab681f8f06/include/bout/mesh.hxx#L388-L391
I think the docstrings are wrong and should be changed (or we need to change `firstY()` and `lastY()` to do what they say).
Limiter configurations are still a bit dodgy because we could have processors where some x-indices have a y-boundary and others don't, but I don't think it's worth worrying about whether to and how to support this case until somebody actually wants it and can make the implementation support their use case.

xBOUT doesn't support `FieldPerp` at all at the moment. We should update that some time after this issue has defined the correct behaviour.
https://github.com/boutproject/xBOUT/blob/1f2090e0167b5147a0489889151a76c625b47fc9/xbout/load.py#L700-L707

Reply:
One problem with allowing saving `FieldPerp`s in guard cells is that at high enough core counts, all of the interior points are also guard cells on neighbouring processors, so it's inevitable that a `FieldPerp` would get written by two processors. This is exactly the problem with `test-io`.

Options:

1. Allow saving in guard cells, and `collect` either just uses the first one it comes across (and assumes that guard & grid cell values agree, possibly checks), or works out which one is a grid cell and uses that
2. Only allow reading/writing to grid & boundary cells
3. Only allow `FieldPerp`s to exist on grid & boundary cells (involves changes to `Datafile::read_fperp` and `GridFile::readField`)

The last option feels the "cleanest" in some sense, but I don't know in what circumstances someone would want a `FieldPerp` in guard cells and therefore when this would break someone's code. `Laplacian` has a `include_yguards` option, but I think that's just a misnomer, and should be `include_y_boundaries`.

The first option feels the most pragmatic, but I don't think it's going to be very easy for `collect` work out if a cell is a boundary cell or a guard cell. We could of course write more information about the mesh/processor decomposition to the restart/dump files, but then we need to worry about how xBOUT or `restart.redistribute` handles that information.

So, the second option then? Actually, this still writes guard cells, it's just that we'd set the `yindex_global` attribute to -1. `collect` wouldn't read it, but it would be possible to read through the python `DataFile` or lower level APIs. The reverse is not true though, and it would not be possible to read guard cells.

Reply:
Thanks @ZedThree. I agree your option 2 is the way to go. :+1:

For the record, I think what I wrote above still wasn't quite right. My actual use case is that I have some sheath boundary conditions that create `FieldPerp`s, and save those to output and restart files. So I have a field like `psi_sheath_lower` that is created and saved anywhere there is a 'lower' sheath boundary. In double-null there are two of those, so I do save `psi_sheath_lower` twice with different `yindex_global` values (which contradicts what I said should be allowed earlier...). Because of the way we handle topology, the two locations where I save `psi_sheath_lower` can never be on the same processor, so restarting works fine.

I'll make an issue on xBOUT about this now. I think I know what we want to implement...